File: commands/stats/topclans.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone
from collections import Counter
from database import db

class TopClansView(discord.ui.View):
    def __init__(self, all_wins, contest_wins, page=0, contest_mode=False):
        super().__init__(timeout=300)
        self.all_wins = all_wins
        self.contest_wins = contest_wins
        self.page = page
        self.contest_mode = contest_mode
        self.per_page = 10
        
        current_data = contest_wins if contest_mode else all_wins
        self.max_pages = (len(current_data) - 1) // self.per_page + 1 if current_data else 1
        
        if self.page <= 0:
            self.previous_button.disabled = True
        if self.page >= self.max_pages - 1:
            self.next_button.disabled = True
            
        self.contest_button.style = discord.ButtonStyle.primary if contest_mode else discord.ButtonStyle.secondary
        self.all_button.style = discord.ButtonStyle.secondary if contest_mode else discord.ButtonStyle.primary

    @discord.ui.button(label='All Wins', style=discord.ButtonStyle.primary, row=0)
    async def all_button(self, interaction: discord.Interaction, button: discord.ui.Button):
        if self.contest_mode:
            self.contest_mode = False
            if self.page >= (len(self.all_wins) - 1) // self.per_page + 1:
                self.page = max(0, (len(self.all_wins) - 1) // self.per_page)
            await interaction.response.edit_message(embed=self.create_embed(), view=TopClansView(self.all_wins, self.contest_wins, self.page, False))

    @discord.ui.button(label='Contest Only', style=discord.ButtonStyle.secondary, row=0)
    async def contest_button(self, interaction: discord.Interaction, button: discord.ui.Button):
        if not self.contest_mode:
            self.contest_mode = True
            if self.page >= (len(self.contest_wins) - 1) // self.per_page + 1:
                self.page = max(0, (len(self.contest_wins) - 1) // self.per_page)
            await interaction.response.edit_message(embed=self.create_embed(), view=TopClansView(self.all_wins, self.contest_wins, self.page, True))

    @discord.ui.button(label='◀ Previous', style=discord.ButtonStyle.secondary, row=1)
    async def previous_button(self, interaction: discord.Interaction, button: discord.ui.Button):
        if self.page > 0:
            self.page -= 1
            await interaction.response.edit_message(embed=self.create_embed(), view=TopClansView(self.all_wins, self.contest_wins, self.page, self.contest_mode))

    @discord.ui.button(label='Next ▶', style=discord.ButtonStyle.secondary, row=1)
    async def next_button(self, interaction: discord.Interaction, button: discord.ui.Button):
        if self.page < self.max_pages - 1:
            self.page += 1
            await interaction.response.edit_message(embed=self.create_embed(), view=TopClansView(self.all_wins, self.contest_wins, self.page, self.contest_mode))

    def create_embed(self):
        current_data = self.contest_wins if self.contest_mode else self.all_wins
        start = self.page * self.per_page
        end = start + self.per_page
        page_data = current_data[start:end]
        
        title = "Top Clans - Contest" if self.contest_mode else "Top Clans - Total"
        description = "Ranked by contest wins" if self.contest_mode else "Ranked by total wins"
        
        embed = discord.Embed(title=title, description=description, color=0x2b2d31)
        
        clan_list = []
        for i, (clan_name, wins) in enumerate(page_data, start + 1):
            medal = "#1" if i == 1 else "#2" if i == 2 else "#3" if i == 3 else f"#{i}"
            clan_list.append(f"{medal} [{clan_name}] - {wins}")
        
        embed.add_field(name="Rankings", value="\n".join(clan_list) if clan_list else "No data", inline=False)
        embed.set_footer(text=f"Data from 18 Nov 2025 | Page {self.page + 1}/{self.max_pages}")
        
        return embed

class TopClans(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='topclans', description='Show top clans by total wins')
    async def topclans(self, interaction: discord.Interaction, days: int = None):
        try:
            await interaction.response.defer()
            
            if days is not None and days < 0:
                embed = discord.Embed(title="❌ Invalid Input", description="Days must be 0 or greater (0 = last 24 hours, 1 = last 2 days, etc.)", color=0xff0000)
                await interaction.followup.send(embed=embed, ephemeral=True)
                return
            
            match_stage = {}
            if days is not None:
                from datetime import timedelta
                now = datetime.now(timezone.utc)
                hours_back = (days + 1) * 24
                cutoff_date = now - timedelta(hours=hours_back)
                match_stage = {'timestamp': {'$gte': cutoff_date}}
            
            pipeline_all = [
                {'$match': match_stage} if match_stage else {'$match': {}},
                {'$group': {'_id': '$clan_name', 'count': {'$sum': 1}}},
                {'$sort': {'count': -1}}
            ]
            
            match_contest = match_stage.copy() if match_stage else {}
            match_contest['contest'] = True
            pipeline_contest = [
                {'$match': match_contest},
                {'$group': {'_id': '$clan_name', 'count': {'$sum': 1}}},
                {'$sort': {'count': -1}}
            ]
            
            all_results = await db.db.clanwins.aggregate(pipeline_all).to_list(None)
            contest_results = await db.db.clanwins.aggregate(pipeline_contest).to_list(None)
            
            all_clans = [(r['_id'], r['count']) for r in all_results if r.get('_id')]
            contest_clans = [(r['_id'], r['count']) for r in contest_results if r.get('_id')]
            
            if not all_clans:
                embed = discord.Embed(
                    title="🏆 Top Clans by Wins",
                    description="No valid clan data found.",
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            view = TopClansView(all_clans, contest_clans)
            embed = view.create_embed()
            await interaction.followup.send(embed=embed, view=view)
            
        except Exception as e:
            print(f"Error in topclans command: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to fetch clan rankings.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(TopClans(bot))

File: commands/stats/topclans.js
// https://github.com/viktorexe/terristats-discord-bot
// /topclans - top clans by total wins, with buttons to toggle between all
// wins and contest-only wins plus Previous/Next pagination.
const {
  SlashCommandBuilder,
  EmbedBuilder,
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle,
  ComponentType,
} = require('discord.js');
const db = require('../../database');

const PER_PAGE = 10;

module.exports = {
  data: new SlashCommandBuilder()
    .setName('topclans')
    .setDescription('Show top clans by total wins')
    .addIntegerOption((o) =>
      o.setName('days').setDescription('Optional day filter').setRequired(false),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const days = interaction.options.getInteger('days');
      if (days !== null && days < 0) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('❌ Invalid Input')
              .setDescription(
                'Days must be 0 or greater (0 = last 24 hours, 1 = last 2 days, etc.)',
              )
              .setColor(0xff0000),
          ],
        });
        return;
      }

      let matchStage = {};
      if (days !== null) {
        const hoursBack = (days + 1) * 24;
        matchStage = { timestamp: { $gte: new Date(Date.now() - hoursBack * 3600 * 1000) } };
      }

      const pipelineAll = [
        { $match: matchStage },
        { $group: { _id: '$clan_name', count: { $sum: 1 } } },
        { $sort: { count: -1 } },
      ];
      const pipelineContest = [
        { $match: { ...matchStage, contest: true } },
        { $group: { _id: '$clan_name', count: { $sum: 1 } } },
        { $sort: { count: -1 } },
      ];

      const [allResults, contestResults] = await Promise.all([
        db.collection('clanwins').aggregate(pipelineAll).toArray(),
        db.collection('clanwins').aggregate(pipelineContest).toArray(),
      ]);

      const allClans = allResults.filter((r) => r._id).map((r) => [r._id, r.count]);
      const contestClans = contestResults.filter((r) => r._id).map((r) => [r._id, r.count]);

      if (!allClans.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('🏆 Top Clans by Wins')
              .setDescription('No valid clan data found.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      let page = 0;
      let contestMode = false;

      const currentData = () => (contestMode ? contestClans : allClans);
      const maxPages = () => Math.max(1, Math.ceil(currentData().length / PER_PAGE));

      const buildEmbed = () => {
        const data = currentData();
        const start = page * PER_PAGE;
        const pageData = data.slice(start, start + PER_PAGE);
        const title = contestMode ? 'Top Clans - Contest' : 'Top Clans - Total';
        const description = contestMode ? 'Ranked by contest wins' : 'Ranked by total wins';
        const clanList = pageData
          .map(([clan, wins], i) => `#${start + i + 1} [${clan}] - ${wins}`)
          .join('\n');
        return new EmbedBuilder()
          .setTitle(title)
          .setDescription(description)
          .setColor(0x2b2d31)
          .addFields({ name: 'Rankings', value: clanList || 'No data', inline: false })
          .setFooter({ text: `Data from 18 Nov 2025 | Page ${page + 1}/${maxPages()}` });
      };

      const buildRows = () => {
        const modeRow = new ActionRowBuilder().addComponents(
          new ButtonBuilder()
            .setCustomId('all')
            .setLabel('All Wins')
            .setStyle(contestMode ? ButtonStyle.Secondary : ButtonStyle.Primary),
          new ButtonBuilder()
            .setCustomId('contest')
            .setLabel('Contest Only')
            .setStyle(contestMode ? ButtonStyle.Primary : ButtonStyle.Secondary),
        );
        const navRow = new ActionRowBuilder().addComponents(
          new ButtonBuilder()
            .setCustomId('prev')
            .setLabel('◀ Previous')
            .setStyle(ButtonStyle.Secondary)
            .setDisabled(page <= 0),
          new ButtonBuilder()
            .setCustomId('next')
            .setLabel('Next ▶')
            .setStyle(ButtonStyle.Secondary)
            .setDisabled(page >= maxPages() - 1),
        );
        return [modeRow, navRow];
      };

      await interaction.editReply({ embeds: [buildEmbed()], components: buildRows() });
      const message = await interaction.fetchReply();

      const collector = message.createMessageComponentCollector({
        componentType: ComponentType.Button,
        time: 300000,
      });

      collector.on('collect', async (i) => {
        if (i.customId === 'all') {
          contestMode = false;
          if (page >= maxPages()) page = Math.max(0, maxPages() - 1);
        } else if (i.customId === 'contest') {
          contestMode = true;
          if (page >= maxPages()) page = Math.max(0, maxPages() - 1);
        } else if (i.customId === 'prev' && page > 0) {
          page -= 1;
        } else if (i.customId === 'next' && page < maxPages() - 1) {
          page += 1;
        }
        await i.update({ embeds: [buildEmbed()], components: buildRows() });
      });

      collector.on('end', async () => {
        try {
          await message.edit({ components: [] });
        } catch (e) {
          // ignore
        }
      });
    } catch (err) {
      console.log(`Error in topclans command: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to fetch clan rankings.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
